src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Android.cs

```diff
@@ -0,0 +1,390 @@
+        */
+        private sealed class AndroidTzData
+        {
+            [StructLayout(LayoutKind.Sequential, Pack=1)]
+            private unsafe struct AndroidTzDataHeader
+            {
+                public fixed byte signature[12]; // "tzdata2012f\0"
+                public int indexOffset;
+                public int dataOffset;
+                public int finalOffset;
+            }
+
+            [StructLayout(LayoutKind.Sequential, Pack=1)]
+            private unsafe struct AndroidTzDataEntry
+            {
+                public fixed byte id[40];
+                public int byteOffset;
+                public int length;
+                public int unused; // Was raw GMT offset; always 0 since tzdata2014f (L).
+            }
+
+            private string[]? _ids;
+            private int[]? _byteOffsets;
+            private int[]? _lengths;
+
+            public AndroidTzData()
+            {
+                ReadHeader(GetTimeZoneDirectory() + _timeZoneFileName);
+            }
+
+            private unsafe void ReadHeader(string tzFilePath)
+            {
+                int size = Math.Max(Marshal.SizeOf(typeof(AndroidTzDataHeader)), Marshal.SizeOf(typeof(AndroidTzDataEntry)));
+                var buffer = new byte[size];
+                AndroidTzDataHeader header = ReadAt<AndroidTzDataHeader>(tzFilePath, 0, buffer);
+
+                header.indexOffset = NetworkToHostOrder(header.indexOffset);
+                header.dataOffset = NetworkToHostOrder(header.dataOffset);
+
+                // tzdata files are expected to start with the form of "tzdata2012f\0" depending on the year of the tzdata used which is 2012 in this example
+                sbyte* s = (sbyte*)header.signature;
+                string magic = new string(s, 0, 6, Encoding.ASCII);
+
+                if (magic != "tzdata" || header.signature[11] != 0)
+                {
+                    var b = new StringBuilder();
+                    for (int i = 0; i < 12; ++i) {
+                        b.Append(' ').Append(((byte)s[i]).ToString("x2"));
+                    }
+
+                    throw new InvalidOperationException(SR.Format(SR.InvalidOperation_BadTZHeader, tzFilePath, b.ToString()));
+                }
+
+                ReadIndex(tzFilePath, header.indexOffset, header.dataOffset, buffer);
+            }
+
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2087:UnrecognizedReflectionPattern",
+                Justification = "Implementation detail of Android TimeZone")]
+            private unsafe T ReadAt<T>(string tzFilePath, long position, byte[] buffer)
+                where T : struct
+            {
+                int size = Marshal.SizeOf(typeof(T));
+                if (buffer.Length < size)
+                {
+                    throw new InvalidOperationException(SR.InvalidOperation_BadBuffer);
+                }
+
+                using (FileStream fs = File.OpenRead(tzFilePath))
```

Comment:
Why are we opening this file multiple times? I see ReadAt is called twice from ReadHeader, and we open the file each time.

Reply:
https://github.com/dotnet/runtime/pull/54845#discussion_r663017991
Originally, the implementation was like https://github.com/mono/mono/blob/e44d84d10a0a6bfe8376cec30582c5c93fd3a92f/mcs/class/corlib/System/TimeZoneInfo.Android.cs#L76 where there was a field that would hold the stream, and would be locked/read when appropriate. Looking at it again, would it be preferred to instead of having a field, just open the stream in the constructor and pass it through instead.

Reply:
looking again, I see this is called in a loop. Yes, just opening and closing the file once in the ctor makes sense to me.

Reply:
Done!